File: 4008-minimum-initial-strength-to-defeat-all-monsters/4008-minimum-initial-strength-to-defeat-all-monsters.cpp
class Solution {
public:
    using ll = long long;

    long long minInitialStrength(vector<int>& nums, vector<vector<int>>& boosts) {
        int n = nums.size();
        vector<ll> diff(n+1, 0);
        for(auto& it : boosts){
            diff[it[0]] += it[2];
            diff[it[1]+1] -= it[2];
        }
        for(int i=1; i<n; i++){
            diff[i]+=diff[i-1];
        }
        ll lo = 0, hi = accumulate(nums.begin(), nums.end(), 0LL);
        ll ans = 0;
        while(lo<=hi){
            ll mid = lo + (hi-lo)/2;
            if(valid(nums,diff, mid)){
                ans = mid;
                hi = mid-1;
            }else lo = mid+1;
        }
        return ans;
    }
    bool valid(vector<int>& nums,vector<ll>& diff, ll s){
        for(int i =0; i<nums.size(); i++){
            int x = nums[i];
            if(x>s){
                if(x>s+diff[i]) return false;
            }
            s-=x;
            if(s<0) s=0;
        }
        return true;
    }
};

File: 4008-minimum-initial-strength-to-defeat-all-monsters/4008-minimum-initial-strength-to-defeat-all-monsters.java
class Solution {
    public long minInitialStrength(int[] monsters, int[][] boosts) {
        int n = monsters.length;

        long[] diff = new long[(int)(n + 1)];

        for(int i = 0 ; i < boosts.length ; i++){
            int first = boosts[i][0];
            int second = boosts[i][1];

            long boost = boosts[i][2];

            diff[first] += boost;

            if(second + 1 <= n)
                diff[second + 1] -= boost;
            
        }

        for(int i = 1 ; i < n ; i++){
            diff[i] = diff[i] + diff[i - 1];
        }
        
        long low = 0;
        long high = (long)1e18;
        long ans = 0;

        while(low <= high){
            long mid = low + (high - low) / 2;

            if(isPossible(monsters, mid, diff)){
                ans = mid;
                high = mid - 1;
            }else{
                low = mid + 1;
            }
        }
        return ans;
    }

    public boolean isPossible(int[] arr, long mid, long[] diff){
        int n = arr.length;

        for(int i = 0 ; i < arr.length ; i++){
            if(arr[i] > mid + diff[i])
            return false;

            mid = mid - arr[i];

            if(mid < 0) mid = 0;
        }

        return true;
    }
}